File: gomoku_9/gomoku_play.py
import torch
import torch.optim as optim
import numpy as np

from alphazero import AlphaZero
from gomoku import Gomoku
from nets import ResNet
from utils import print_board

if __name__ == '__main__':
    np.set_printoptions(precision=2, suppress=True)
    game = Gomoku(board_size=9, history_step=4)
    model = ResNet(game, num_blocks=4, num_channels=256, history_step=4).to('cuda')
    optimizer = None
    args = {
        'mode': 'eval',
        'num_simulations': 600,
        'c_puct': 1,
        'buffer_size': 100000,
        'file_name': 'gomoku',
        'device': 'cuda'
    }

    alphazero = AlphaZero(game, model, optimizer, args)
    alphazero.load_checkpoint()

    to_play = int(input(
        f'1 for the first move and -1 for the second move\n'
        f'The position of the piece needs to be input in coordinate form.\n'
        f'   (first input the vertical coordinate, then the horizontal coordinate).\n'
        f'Please enter:'
    ))
    color = 1
    state = game.get_initial_state()
    print_board(state)
    while not game.is_terminal(state):
        if to_play == 1:
            move = input(f"Human step:")
            i, j = map(int, move.strip().split())
            action = i * game.board_size + j
            state = game.get_next_state(state, action, color)
        elif to_play == -1:
            print(f'AlphaZero step:')
            action, info = alphazero.play(state, color)
            state = game.get_next_state(state, action, color)
            action_probs, policy, ai_winrate = info['action_probs'], info['policy'], info['ai_winrate']
            print(action_probs.reshape(game.board_size, game.board_size))
            print(policy)
            print(ai_winrate)
        to_play = -to_play
        color = -color
        print_board(state)

File: CSkyZero_V3/gomoku/gomoku_play.cpp
#include <iostream>

#include <torch/torch.h>

#include "../envs/gomoku.h"
#include "../playgame.h"

namespace {

skyzero::AlphaZeroConfig build_gomoku_eval_config() {
    skyzero::AlphaZeroConfig cfg;
    cfg.board_size = 15;
    cfg.num_blocks = 4;
    cfg.num_channels = 128;
    cfg.num_simulations = 400;
    cfg.device = torch::cuda::is_available() ? torch::kCUDA : torch::kCPU;
    return cfg;
}

}  // namespace

int main() {
    auto cfg = build_gomoku_eval_config();
    cfg.lr = 1e-4f;
    cfg.weight_decay = 3e-5f;
    cfg.file_name = "gomoku";
    cfg.data_dir = "data/gomoku";

    skyzero::Gomoku game(cfg.board_size, true, true);
    skyzero::GamePlayer<skyzero::Gomoku> player(game, cfg);
    player.load_checkpoint();
    return player.play();
}
